Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
+    }
+    return value;
+}
+
+
+- (NSDictionary*) currentPropertiesAsJson: (BOOL)asJson {
+    NSMutableDictionary* result;
+    NSMutableSet *changedNames = [_changedNames mutableCopy];
+    if (_root) {
+        int count = FLDict_Count(_root);
+        result = [NSMutableDictionary dictionaryWithCapacity: count];
+        if (count > 0) {
+            FLDictIterator iter;
+            FLDictIterator_Begin(_root, &iter);
+            do {
+                NSString* key = [self fleeceValueToKeyString: FLDictIterator_GetKey(&iter)];
+                id value = _properties[key];
+                if (value) {
+                    if (asJson)
+                        value = [self jsonValueForValue: value];
+                    [changedNames removeObject: key];
+                } else {
+                    value = [self fleeceValueToObject: FLDictIterator_GetValue(&iter)
+                                               forKey: key asJson: asJson];
+                    [self cacheValue: value forKey: key changed: NO];
+                }
+                result[key] = value;
+            } while (FLDictIterator_Next(&iter));
+        }
+        return result;
+    }
+    if (changedNames.count > 0) {
+        if (!result)
+            result = [NSMutableDictionary dictionaryWithCapacity: changedNames.count];
+        for (NSString* key in _changedNames) {
+            id value = _properties[key];
+            result[key] = asJson ? [self jsonValueForValue: value] : value;
+        }
+    }
+    return result;
+}
+
+
+- (id) jsonValueForValue: (id)value {
+    if ([value conformsToProtocol: @protocol(CBLJSONEncoding)]) {
+        return [(id<CBLJSONEncoding>)value encodeAsJSON];
+    } else if ([value isKindOfClass: [NSDictionary class]]) {
+        NSDictionary* dict = $cast(NSDictionary, value);
+        NSMutableDictionary* result = [dict mutableCopy];
+        for (NSString* key in dict) {
+            id obj = dict[key];
+            id extObj = [self jsonValueForValue: obj];
+            if (obj != extObj)
+                result[key] = extObj;
+        }
+        return result;
+    } else if ([value isKindOfClass: [NSArray class]]) {
+        NSArray* array = $cast(NSArray, value);
+        NSMutableArray* result = [NSMutableArray arrayWithCapacity: array.count];
+        for (id obj in array) {
+            id extObj = [self jsonValueForValue: obj];
+            if (extObj)
+                [result addObject: extObj];
+        } return result;
+    } else {
+        return value;
+    }
+}
+
+
+#pragma mark - PRIVATE: CACHE
+
+
+- (void) cacheValue: (nullable id)value forKey: (NSString*)key changed: (BOOL)changed {
+    BOOL shouldCache = changed ||
+                       [value isKindOfClass: [CBLSubdocument class]] ||
+                       [value isKindOfClass: [NSArray class]];
+    if (shouldCache) {
+        if (!_properties)
+            _properties = [NSMutableDictionary dictionary];
+        _properties[key] = value;
```

Comment:
This will crash if value is nil.
An alternative is `[_properties setValue: value forKey: key]`, which allows a nil value.